File: lab4/wc.pl
#!/usr/bin/perl -T

use strict;
use warnings qw(FATAL all);

my $wantLines = 1;
my $wantWords = 1;
my $wantBytes = 1;

my $filename = "";
my $i = 0;

my $totalLines = 0;
my $totalWords = 0;
my $totalBytes = 0;
my $fileCounter = 0;

my $lines = 0;
my $words = 0;
my $bytes = 0;
 
if ($ARGV[0] !~ '^$') {
    if ($#ARGV + 1 == 0) {
        print "Incorrect arguments! Use wc -lwc filename."
    } else {
        if ($ARGV[0] =~ '^-') {
                if ($ARGV[0] !~ 'l') {
                    $wantLines = 0;
                }
                if ($ARGV[0] !~ 'c') {
                    $wantBytes = 0;
                }
                if ($ARGV[0] !~ 'w') {
                    $wantWords = 0;
                }
                $i = 1;
        }

        for ($i = $i; $i < $#ARGV + 1; $i++) {
            $filename = $ARGV[$i];
            $fileCounter++;
            if (-e $filename) {
                if (-d $filename) {
                    print "File $filename is a directory!";
                    next; 
                } else {
                    $lines = 0;
                    $words = 0;
                    $bytes = 0;
                    open(my $fh, $filename);
 
                    while (my $row = <$fh>) {
                        if ($wantBytes) {
                            $bytes += length($row);
                        }
                        if ($wantLines) {
                            $lines++;
                        }
                        if ($wantWords) {
                            $words += scalar(split(/\W+/, $row));
                        }
                    }

                    if ($wantLines) { print "$lines "; }
                    if ($wantWords) { print "$words "; }
                    if ($wantBytes) { print "$bytes "; } 
                    print "$filename \n";
                }
            } else {
                print "File $filename doesn't exist!";
                next;
            }

            if ($wantBytes) { $totalBytes += $bytes; }
            if ($wantLines) { $totalLines += $lines; }
            if ($wantWords) { $totalWords += $words; }
        }

        if ($fileCounter > 1) {
            if ($wantLines) { print "$totalLines "; }
            if ($wantWords) { print "$totalWords "; }
            if ($wantBytes) { print "$totalBytes "; } 
            print "total\n";
        }

    }
} else {
    print "Enter file name!";
}

File: lab4/wc.c
#include <stdlib.h>
#include <stdio.h>
#include <stdbool.h>
#include <errno.h>
#include <string.h>
#include <fcntl.h>
#include <sys/stat.h>
#include <unistd.h>

bool wantBytes = true;
bool wantLines = true;
bool wantWords = true;
bool hasFlags = false;

int fileCounter = 0;
int totalLines = 0;
int totalWords = 0;
int totalBytes = 0;

void wc(char* filename) {
    errno = 0;
    int fd = open(filename, O_RDONLY);
    if (fd == -1) {
        fprintf(stderr, "File %s doesn't exist!\n", filename);
        return;
    }
    struct stat s;
    if( stat(filename,&s) == 0 ) {
        if (S_ISDIR(s.st_mode)) {
            fprintf(stderr, "File %s is directory!\n", filename);
            return;
        }
    } else {
        fprintf(stderr, "Exception during opening file %s!\n", filename);
        return;
    }

    fileCounter++;

    int lines = 0;
    int words = 0;
    int bytes = 0;

    char *buf = (char*)malloc(1024);
    char *prevBuf = (char*)malloc(1024);

    while (read(fd, buf, sizeof(*buf))) {
        if (wantBytes) {
            bytes++;
        }
        if (wantLines) {
            if (strcmp(buf, "\n") == 0) {
                lines++;
            }
        }
        if (wantWords) {
            if (((strcmp(buf, " ") == 0) || (strcmp(buf, "\n") == 0) || (strcmp(buf, "\t") == 0)) && (strcmp(buf, prevBuf) != 0)) {
                words++;
            }
        }
        *prevBuf = *buf;
    }

    close(fd);

    totalWords += words;
    totalLines += lines;
    totalBytes += bytes;


    if (wantLines) printf("%d ", lines);
    if (wantWords) printf("%d ", words);
    if (wantBytes) printf("%d ", bytes);
    printf("%s\n", filename);
}


int main(int argc, char *argv[]) {
    if (argc < 2) {
        fprintf(stderr, "Incorrect number of arguments! Usage wc -lwc filename\n");
        return 0;
    }

    if ((argc > 2) && (strstr(argv[1],"-") != NULL) && ((strstr(argv[1],"l") != NULL) || (strstr(argv[1],"w") != NULL) || (strstr(argv[1],"c") != NULL))) {
            hasFlags = true;
            if (strstr(argv[1], "l") == NULL) {
                wantLines = false;
            }
            if (strstr(argv[1], "w") == NULL) {
                wantWords = false;
            }
            if (strstr(argv[1], "c") == NULL) {
                wantBytes = false;
            }

    } else {
        if ((argc > 2) && (strstr(argv[1],"-") != NULL)) {
            fprintf(stderr, "Incorrect arguments! Usage wc -lwc filename\n");
            return 0;
        }
    }

    for (int i = hasFlags ? 2 : 1; i < argc; i++) {
        wc(argv[i]);
    }

    if (fileCounter > 1) {
        if (wantLines) printf("%d ", totalLines);
        if (wantWords) printf("%d ", totalWords);
        if (wantBytes) printf("%d ", totalBytes);
        printf("total\n");
    }

    return 0;
}


